<a href="https://colab.research.google.com/github/Jon227student/IS_170_Machine_Learning/blob/5.-Association-Rules/HW_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

HW#5 Association rules

Gaming store sales

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
df=pd.read_csv('/content/new data.csv')
df.head()

racegame,shootergame,controller
0   console,controller,candy,drink,shootergame
1             drink,racegame,shootergame,candy
2  gaminglaptop,drink,strategygame,candy,mouse
3                   strategygame,rpggame,mouse
4                    console,controller,drink

In [ ]:
df=pd.read_csv('/content/new data.csv', names=['sales'])
df.head() 


sales
0              racegame,shootergame,controller
1   console,controller,candy,drink,shootergame
2             drink,racegame,shootergame,candy
3  gaminglaptop,drink,strategygame,candy,mouse
4                   strategygame,rpggame,mouse

In [ ]:
df.shape

(14, 1)

In [ ]:
df=pd.read_csv('/content/new data.csv', names=['sales'])
df.head() 

sales
0              racegame,shootergame,controller
1   console,controller,candy,drink,shootergame
2             drink,racegame,shootergame,candy
3  gaminglaptop,drink,strategygame,candy,mouse
4                   strategygame,rpggame,mouse

In [ ]:
data=list(df["sales"].apply(lambda x:x.split(",")))
data

[['racegame', 'shootergame', 'controller'],
 ['console', 'controller', 'candy', 'drink', 'shootergame'],
 ['drink', 'racegame', 'shootergame', 'candy'],
 ['gaminglaptop', 'drink', 'strategygame', 'candy', 'mouse'],
 ['strategygame', 'rpggame', 'mouse'],
 ['console', 'controller', 'drink '],
 ['mouse', 'strategygame', 'racegame'],
 ['rpggame', 'strategygame', 'gaminglaptop'],
 ['console', 'shooter', 'gamecandy'],
 ['rpggame', 'strategygame', 'shootergame '],
 ['shootergame', 'drink', 'candy'],
 ['rpggame', 'gaminglaptop', 'mouse', 'strategygame  '],
 ['shootergame', 'candy', 'racegame'],
 ['gaminglaptop', 'mouse', 'candy', 'rpggame ']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

df=df.replace(False,0)
df=df.replace(True,1)

df
                

candy  console  controller  drink  drink   gamecandy  gaminglaptop  mouse  \
0       0        0           1      0       0          0             0      0   
1       1        1           1      1       0          0             0      0   
2       1        0           0      1       0          0             0      0   
3       1        0           0      1       0          0             1      1   
4       0        0           0      0       0          0             0      1   
5       0        1           1      0       1          0             0      0   
6       0        0           0      0       0          0             0      1   
7       0        0           0      0       0          0             1      0   
8       0        1           0      0       0          1             0      0   
9       0        0           0      0       0          0             0      0   
10      1        0           0      1       0          0             0      0   
11      0        0           0      0       0          0             1      1   
12      1        0           0      0       0          0             0      0   
13      1        0           0      0       0          0             1      1   

    racegame  rpggame  rpggame   shooter  shootergame  shootergame   \
0          1        0         0        0            1             0   
1          0        0         0        0            1             0   
2          1        0         0        0            1             0   
3          0        0         0        0            0             0   
4          0        1         0        0            0             0   
5          0        0         0        0            0             0   
6          1        0         0        0            0             0   
7          0        1         0        0            0             0   
8          0        0         0        1            0             0   
9          0        1         0        0            0             1   
10         0        0         0        0            1             0   
11         0        1         0        0            0             0   
12         1        0         0        0            1             0   
13         0        0         1        0            0             0   

    strategygame  strategygame    
0              0               0  
1              0               0  
2              0               0  
3              1               0  
4              1               0  
5              0               0  
6              1               0  
7              1               0  
8              0               0  
9              1               0  
10             0               0  
11             0               1  
12             0               0  
13             0               0

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support                     itemsets
0   0.428571                      (candy)
1   0.214286                    (console)
2   0.214286                 (controller)
3   0.285714                      (drink)
4   0.285714               (gaminglaptop)
5   0.357143                      (mouse)
6   0.285714                   (racegame)
7   0.285714                    (rpggame)
8   0.357143                (shootergame)
9   0.357143               (strategygame)
10  0.285714               (drink, candy)
11  0.285714         (shootergame, candy)
12  0.214286         (drink, shootergame)
13  0.214286        (mouse, gaminglaptop)
14  0.214286        (strategygame, mouse)
15  0.214286      (racegame, shootergame)
16  0.214286      (strategygame, rpggame)
17  0.214286  (shootergame, drink, candy)

The above values show our most frequently purchased items it seems candy is our most purchased item with 42% of our customers buying candy in their purchases. The next two are the new shooter game and the new strategy game being our top sellers, 35% of our customers include these in their purchases

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents           consequents  antecedent support  \
0                (drink)               (candy)            0.285714   
1                (candy)               (drink)            0.428571   
2          (shootergame)               (candy)            0.357143   
3                (candy)         (shootergame)            0.428571   
4                (drink)         (shootergame)            0.285714   
5          (shootergame)               (drink)            0.357143   
6                (mouse)        (gaminglaptop)            0.357143   
7         (gaminglaptop)               (mouse)            0.285714   
8         (strategygame)               (mouse)            0.357143   
9                (mouse)        (strategygame)            0.357143   
10            (racegame)         (shootergame)            0.285714   
11         (shootergame)            (racegame)            0.357143   
12        (strategygame)             (rpggame)            0.357143   
13             (rpggame)        (strategygame)            0.285714   
14  (drink, shootergame)               (candy)            0.214286   
15  (candy, shootergame)               (drink)            0.285714   
16        (drink, candy)         (shootergame)            0.285714   
17         (shootergame)        (drink, candy)            0.357143   
18               (drink)  (candy, shootergame)            0.285714   
19               (candy)  (drink, shootergame)            0.428571   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.428571  0.285714    1.000000  2.333333  0.163265         inf  
1             0.285714  0.285714    0.666667  2.333333  0.163265    2.142857  
2             0.428571  0.285714    0.800000  1.866667  0.132653    2.857143  
3             0.357143  0.285714    0.666667  1.866667  0.132653    1.928571  
4             0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
5             0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
6             0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
7             0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
8             0.357143  0.214286    0.600000  1.680000  0.086735    1.607143  
9             0.357143  0.214286    0.600000  1.680000  0.086735    1.607143  
10            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
11            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
12            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
13            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
14            0.428571  0.214286    1.000000  2.333333  0.122449         inf  
15            0.285714  0.214286    0.750000  2.625000  0.132653    2.857143  
16            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
17            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
18            0.285714  0.214286    0.750000  2.625000  0.132653    2.857143  
19            0.214286  0.214286    0.500000  2.333333  0.122449    1.571429

We can note from this data that people who buy a candy usually buy a drink along with it 42% of the time also people who buy gaming laptops usually buy strategy games or rpg games 35% of the time

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents           consequents  antecedent support  \
0                (drink)               (candy)            0.285714   
1                (candy)               (drink)            0.428571   
2          (shootergame)               (candy)            0.357143   
3                (candy)         (shootergame)            0.428571   
4                (drink)         (shootergame)            0.285714   
5          (shootergame)               (drink)            0.357143   
6                (mouse)        (gaminglaptop)            0.357143   
7         (gaminglaptop)               (mouse)            0.285714   
8         (strategygame)               (mouse)            0.357143   
9                (mouse)        (strategygame)            0.357143   
10            (racegame)         (shootergame)            0.285714   
11         (shootergame)            (racegame)            0.357143   
12        (strategygame)             (rpggame)            0.357143   
13             (rpggame)        (strategygame)            0.285714   
14  (drink, shootergame)               (candy)            0.214286   
15  (candy, shootergame)               (drink)            0.285714   
16        (drink, candy)         (shootergame)            0.285714   
17         (shootergame)        (drink, candy)            0.357143   
18               (drink)  (candy, shootergame)            0.285714   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.428571  0.285714    1.000000  2.333333  0.163265         inf  
1             0.285714  0.285714    0.666667  2.333333  0.163265    2.142857  
2             0.428571  0.285714    0.800000  1.866667  0.132653    2.857143  
3             0.357143  0.285714    0.666667  1.866667  0.132653    1.928571  
4             0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
5             0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
6             0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
7             0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
8             0.357143  0.214286    0.600000  1.680000  0.086735    1.607143  
9             0.357143  0.214286    0.600000  1.680000  0.086735    1.607143  
10            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
11            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
12            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
13            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
14            0.428571  0.214286    1.000000  2.333333  0.122449         inf  
15            0.285714  0.214286    0.750000  2.625000  0.132653    2.857143  
16            0.357143  0.214286    0.750000  2.100000  0.112245    2.571429  
17            0.285714  0.214286    0.600000  2.100000  0.112245    1.785714  
18            0.285714  0.214286    0.750000  2.625000  0.132653    2.857143

We can see a strong positive correlation between laptop purchasers and strategy games/rpg games. Console purchasers have a positive correlation for shooter games and racing games.Drink and candy also have a high correlation rate

Hey boss looking over this data makes me realize we should improve the variety of our candy section since it is very popular with our customers. It is also clear that laptop purchasers prefer strategy/rpg games while console purchasers enjoy shooter/racing games. We should relocate these games near eachother. 

Sporting goods sales

In [ ]:
df=pd.read_csv('/content/Newdata2.csv')
df.head()

tennis racket,tennis balls,headband,water bottle
0  baseball bat,baseball glove,baseball hat,baseb...
1          basketball,sneakers,water bottle,headband
2   tennis racket,tennis balls,headband,water bottle
3  baseball bat,baseball glove,baseball hat,baseb...
4          basketball,sneakers,water bottle,headband

In [ ]:
df=pd.read_csv('/content/Newdata2.csv', names=['sport goods'])
df.head() 

sport goods
0   tennis racket,tennis balls,headband,water bottle
1  baseball bat,baseball glove,baseball hat,baseb...
2          basketball,sneakers,water bottle,headband
3   tennis racket,tennis balls,headband,water bottle
4  baseball bat,baseball glove,baseball hat,baseb...

In [ ]:
df.shape

(10, 1)

In [ ]:
data=list(df["sport goods"].apply(lambda x:x.split(",")))
data

[['tennis racket', 'tennis balls', 'headband', 'water bottle'],
 ['baseball bat',
  'baseball glove',
  'baseball hat',
  'baseball',
  'water bottle'],
 ['basketball', 'sneakers', 'water bottle', 'headband'],
 ['tennis racket', 'tennis balls', 'headband', 'water bottle'],
 ['baseball bat',
  'baseball glove',
  'baseball hat',
  'baseball',
  'water bottle'],
 ['basketball', 'sneakers', 'water bottle', 'headband'],
 ['tennis racket', 'tennis balls', 'headband', 'water bottle'],
 ['basketball', 'sneakers', 'water bottle', 'headband'],
 ['baseball bat',
  'baseball glove',
  'baseball hat',
  'baseball',
  'water bottle'],
 ['tennis racket', 'tennis balls', 'headband', 'water bottle']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

df=df.replace(False,0)
df=df.replace(True,1)

df

baseball  baseball bat  baseball glove  baseball hat  basketball  headband  \
0         0             0               0             0           0         1   
1         1             1               1             1           0         0   
2         0             0               0             0           1         1   
3         0             0               0             0           0         1   
4         1             1               1             1           0         0   
5         0             0               0             0           1         1   
6         0             0               0             0           0         1   
7         0             0               0             0           1         1   
8         1             1               1             1           0         0   
9         0             0               0             0           0         1   

   sneakers  tennis balls  tennis racket  water bottle  
0         0             1              1             1  
1         0             0              0             1  
2         1             0              0             1  
3         0             1              1             1  
4         0             0              0             1  
5         1             0              0             1  
6         0             1              1             1  
7         1             0              0             1  
8         0             0              0             1  
9         0             1              1             1

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support                                           itemsets
0       0.3                                         (baseball)
1       0.3                                     (baseball bat)
2       0.3                                   (baseball glove)
3       0.3                                     (baseball hat)
4       0.3                                       (basketball)
5       0.7                                         (headband)
6       0.3                                         (sneakers)
7       0.4                                     (tennis balls)
8       0.4                                    (tennis racket)
9       1.0                                     (water bottle)
10      0.3                           (baseball bat, baseball)
11      0.3                         (baseball glove, baseball)
12      0.3                           (baseball hat, baseball)
13      0.3                           (water bottle, baseball)
14      0.3                     (baseball bat, baseball glove)
15      0.3                       (baseball bat, baseball hat)
16      0.3                       (baseball bat, water bottle)
17      0.3                     (baseball hat, baseball glove)
18      0.3                     (water bottle, baseball glove)
19      0.3                       (baseball hat, water bottle)
20      0.3                             (headband, basketball)
21      0.3                             (sneakers, basketball)
22      0.3                         (water bottle, basketball)
23      0.3                               (headband, sneakers)
24      0.4                           (tennis balls, headband)
25      0.4                          (headband, tennis racket)
26      0.7                           (headband, water bottle)
27      0.3                           (sneakers, water bottle)
28      0.4                      (tennis balls, tennis racket)
29      0.4                       (tennis balls, water bottle)
30      0.4                      (water bottle, tennis racket)
31      0.3           (baseball glove, baseball bat, baseball)
32      0.3             (baseball bat, baseball hat, baseball)
33      0.3             (baseball bat, water bottle, baseball)
34      0.3           (baseball glove, baseball hat, baseball)
35      0.3           (baseball glove, water bottle, baseball)
36      0.3             (baseball hat, water bottle, baseball)
37      0.3       (baseball bat, baseball hat, baseball glove)
38      0.3       (baseball bat, water bottle, baseball glove)
39      0.3         (water bottle, baseball bat, baseball hat)
40      0.3       (baseball hat, water bottle, baseball glove)
41      0.3                   (headband, sneakers, basketball)
42      0.3               (headband, water bottle, basketball)
43      0.3               (sneakers, water bottle, basketball)
44      0.3                 (sneakers, headband, water bottle)
45      0.4            (tennis balls, headband, tennis racket)
46      0.4             (tennis balls, headband, water bottle)
47      0.4            (water bottle, headband, tennis racket)
48      0.4        (water bottle, tennis balls, tennis racket)
49      0.3  (baseball glove, baseball bat, baseball hat, b...
50      0.3  (baseball glove, baseball bat, water bottle, b...
51      0.3  (water bottle, baseball bat, baseball hat, bas...
52      0.3  (baseball hat, baseball glove, water bottle, b...
53      0.3  (water bottle, baseball bat, baseball hat, bas...
54      0.3     (sneakers, headband, water bottle, basketball)
55      0.4  (water bottle, headband, tennis balls, tennis ...
56      0.3  (baseball bat, water bottle, baseball, basebal...

It seems that head bands and water bottles are our most purchased goods here at the sporting goods store each having a .7 support rate. We can also gage what sport is selling the most product in this case tennis products have a higher frequency than the others

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents                                        consequents  \
0      (baseball bat)                                         (baseball)   
1          (baseball)                                     (baseball bat)   
2    (baseball glove)                                         (baseball)   
3          (baseball)                                   (baseball glove)   
4      (baseball hat)                                         (baseball)   
..                ...                                                ...   
273    (baseball bat)  (baseball hat, baseball glove, water bottle, b...   
274    (water bottle)  (baseball glove, baseball bat, baseball hat, b...   
275        (baseball)  (baseball hat, baseball bat, water bottle, bas...   
276    (baseball hat)  (baseball glove, baseball bat, water bottle, b...   
277  (baseball glove)  (baseball hat, baseball bat, water bottle, bas...   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.3                 0.3      0.3         1.0  3.333333   
1                   0.3                 0.3      0.3         1.0  3.333333   
2                   0.3                 0.3      0.3         1.0  3.333333   
3                   0.3                 0.3      0.3         1.0  3.333333   
4                   0.3                 0.3      0.3         1.0  3.333333   
..                  ...                 ...      ...         ...       ...   
273                 0.3                 0.3      0.3         1.0  3.333333   
274                 1.0                 0.3      0.3         0.3  1.000000   
275                 0.3                 0.3      0.3         1.0  3.333333   
276                 0.3                 0.3      0.3         1.0  3.333333   
277                 0.3                 0.3      0.3         1.0  3.333333   

     leverage  conviction  
0        0.21         inf  
1        0.21         inf  
2        0.21         inf  
3        0.21         inf  
4        0.21         inf  
..        ...         ...  
273      0.21         inf  
274      0.00         1.0  
275      0.21         inf  
276      0.21         inf  
277      0.21         inf  

[278 rows x 9 columns]

It seems that people buy head bands and water bottles most frequently together it may be because these items can go with all sports there is a 38% chance if they buy a water bottle they will buy a headband. The related sporting items also have a  antecedent of .3 for almost all sports

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar

antecedents  \
0                    (baseball bat)   
1                        (baseball)   
2                  (baseball glove)   
3                        (baseball)   
4                    (baseball hat)   
..                              ...   
228  (baseball hat, baseball glove)   
229                  (baseball bat)   
230                      (baseball)   
231                  (baseball hat)   
232                (baseball glove)   

                                           consequents  antecedent support  \
0                                           (baseball)                 0.3   
1                                       (baseball bat)                 0.3   
2                                           (baseball)                 0.3   
3                                     (baseball glove)                 0.3   
4                                           (baseball)                 0.3   
..                                                 ...                 ...   
228             (baseball bat, water bottle, baseball)                 0.3   
229  (baseball hat, baseball glove, water bottle, b...                 0.3   
230  (baseball hat, baseball bat, water bottle, bas...                 0.3   
231  (baseball glove, baseball bat, water bottle, b...                 0.3   
232  (baseball hat, baseball bat, water bottle, bas...                 0.3   

     consequent support  support  confidence      lift  leverage  conviction  
0                   0.3      0.3         1.0  3.333333      0.21         inf  
1                   0.3      0.3         1.0  3.333333      0.21         inf  
2                   0.3      0.3         1.0  3.333333      0.21         inf  
3                   0.3      0.3         1.0  3.333333      0.21         inf  
4                   0.3      0.3         1.0  3.333333      0.21         inf  
..                  ...      ...         ...       ...       ...         ...  
228                 0.3      0.3         1.0  3.333333      0.21         inf  
229                 0.3      0.3         1.0  3.333333      0.21         inf  
230                 0.3      0.3         1.0  3.333333      0.21         inf  
231                 0.3      0.3         1.0  3.333333      0.21         inf  
232                 0.3      0.3         1.0  3.333333      0.21         inf  

[233 rows x 9 columns]

Most of the related sporting items have a high lift meaning there is a positive correlation they are all over 1. Sporting items with no relation have a negative correlation most all being under 1. The average lift score for related sporting items are 3.333

Hey boss it seems sales are great. We should locate headbands to the front next to water bottles these have a high positve correlation. We should also creates sections around the store for each sport since each items belonging to the same sport also have a high positve correlation.